pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -221,53 +230,262 @@ func New(plArgs runtime.Object, fh framework.Handle, fts feature.Features) (fram
+	if !ok {
+		// Shouldn't happen.
+		pl.logger.Error(nil, "unexpected new object in isSchedulableAfterClaimAddOrUpdate", "obj", newObj)
+		return framework.PodMaybeSchedulable
+	}
+
+	usesClaim := false
+	if err := pl.foreachPodResourceClaim(pod, func(_ string, claim *resourcev1alpha2.ResourceClaim) {
+		if claim.UID == modifiedClaim.UID {
+			usesClaim = true
+		}
+	}); err != nil {
+		// This is not an unexpected error: we know that
+		// foreachPodResourceClaim only returns errors for "not
+		// schedulable".
+		pl.logger.V(4).Info("pod is not schedulable", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim), "reason", err.Error())
+		return framework.PodNotAffected
+	}
+
+	if usesClaim {
+		if oldObj == nil {
+			pl.logger.V(4).Info("claim for pod got created", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim))
+			return framework.PodImmediatelySchedulable
+		}
+
+		// Modifications may or may not be relevant. If the entire
+		// status is as before, then something else must have changed
+		// and we don't care. What happens in practice is that the
+		// resource driver adds the finalizer.
+		originalClaim, ok := oldObj.(*resourcev1alpha2.ResourceClaim)
+		if !ok {
+			// Shouldn't happen.
+			pl.logger.Error(nil, "unexpected old object in isSchedulableAfterClaimAddOrUpdate", "obj", oldObj)
+			return framework.PodMaybeSchedulable
+		}
+		if apiequality.Semantic.DeepEqual(&originalClaim.Status, &modifiedClaim.Status) {
+			if loggerV := pl.logger.V(7); loggerV.Enabled() {
+				// Log more information.
+				loggerV.Info("claim for pod got modified where the pod doesn't care", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim), "diff", cmp.Diff(originalClaim, modifiedClaim))
+			} else {
+				pl.logger.V(6).Info("claim for pod got modified where the pod doesn't care", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim))
+			}
+			return framework.PodNotAffected
+		}
+
+		pl.logger.V(4).Info("status of claim for pod got updated", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim))
+		return framework.PodImmediatelySchedulable
+	}
+
+	// This was not the claim the pod was waiting for.
+	pl.logger.V(6).Info("unrelated claim got modified", "pod", klog.KObj(pod), "claim", klog.KObj(modifiedClaim))
+	return framework.PodNotAffected
+}
+
+// isSchedulableAfterPodSchedulingChange is invoked whenever a
+// PodScheduling object is updated or deleted. It checks whether that change
+// made a previously unschedulable pod schedulable (updated) or a new attempt
+// is needed to re-create the object (deleted). It errs on the side of letting
+// a pod scheduling attempt happen.
+func (pl *dynamicResources) isSchedulableAfterPodSchedulingChange(oldObj, newObj interface{}, pod *v1.Pod) framework.SchedulingHint {
+	obj := newObj
+	if obj == nil {
+		// Delete event.
+		obj = oldObj
+	}
+
+	var podScheduling *resourcev1alpha2.PodSchedulingContext
+	switch obj := obj.(type) {
+	case cache.DeletedFinalStateUnknown:
```

Comment:
Would it make sense to move the type check for `cache.DeletedFinalStateUnknown` into the framework?

The advantage is that plugins become simpler. Plugin developers won't make the mistake of forgetting this check (which should be rarely needed, but sometimes can occur). If multiple plugins filter on the same GVK, then the type check only needs to be done once.

The disadvantage is doing this check whether the hint function needs the object or not. For example, `isSchedulableAfterClaimChange` bails out early on deletes. However, this might not matter at all: if `framework.Delete` is set as action type, then at least one plugin seems to be interested in deletions and thus probably also the deleted object.

I'm leaning towards moving this into the framework.